src/core/mount.c

```diff
@@ -1750,9 +1783,54 @@ static int mount_load_proc_self_mountinfo(Manager *m, bool set_flags) {
                 if (!device || !path)
                         continue;
 
+                if (ignore_mount(path, fstype))
+                        continue;
+
+                size = snprintf(mount_entry, mount_entry_size, "%s %s %s %s", device, path, fstype, options);
+
+                if ((size_t)size >= mount_entry_size) {
+                        if (mount_entry)
+                                free(mount_entry);
+                        mount_entry_size = 1024 * (size / 1024 + 1);
+                        mount_entry = new(char, mount_entry_size);
+                        if (!mount_entry) {
+                                mount_entry_size = 0;
+                                return -ENOMEM;
+                        }
+                        size = snprintf(mount_entry, mount_entry_size, "%s %s %s %s", device, path, fstype, options);
+                }
+
+                mu = hashmap_get2(mountinfo_hashmap, mount_entry, (void **)&key);
+                if (mu) {
+                        assert(mu->mountinfo_key == key);
+                        if (mu->state == MOUNT_MOUNTED) {
+                                mu->proc_flags |= MOUNT_PROC_IS_MOUNTED;
+                                log_debug("Mount entry '%s' already known, won't set device nor update mount unit", mount_entry);
```

Comment:
"Won't set device *or* update mount unit"

Reply:
ok

Reply:
> > "Won't set device _or_ update mount unit"
> 
> Umm ... what's the question here?
> 
> The whole point of this is to identify mount units that haven't changed at all and not call the update functions because the mount unit (including the device) hasn't changed. Calling the update functions is a large part of why having to read the entire mount table (since there's no other choice, and many entries haven't changed at all) on every change notification is such a big problem on heavy mount table activity. There's no real in-kernel merging of notifications to speak of, since if user space can process the notification fairly quickly (and systemd seems fairly good at this) it will see there's been a change on the very next check and on it goes ...
> 
> The only other aspect of this is managing the rate of notifications when the rate of notifications is high and that's a separate problem.

Oh wait, it's the nor -> or that your pointing out, oops!